File: javascript/perfect-numbers/perfect-numbers.js
export const classify = (n) => {
  if (n <= 0)  throw new Error(
                 'Classification is only possible for natural numbers.'
               );
  let factors = 0;

  for(let f = 1; f < n; f++) {
    if (n % f === 0) {
      factors += f;
    }
  }

  if (factors === n) return 'perfect';
  if (factors > n) return 'abundant';

  return 'deficient';
};

File: ruby/perfect-numbers/perfect_numbers.rb
class PerfectNumber
  def self.classify(num)
    raise RuntimeError if num < 1

    aliquot_sum = 0
    1.upto(num - 1) { |i| aliquot_sum += i if (num % i).zero? }
    return 'abundant' if aliquot_sum > num
    return 'deficient' if aliquot_sum < num

    'perfect'
  end
end
